File: fsharp/interest-is-interesting/InterestIsInteresting.fs
module InterestIsInteresting

let interestRate (balance: decimal): single =
    match balance with
    | balance when balance < 0m -> 3.213f
    | balance when balance < 1_000m -> 0.5f
    | balance when balance < 5_000m -> 1.621f
    | _ -> 2.475f

let interest (balance: decimal): decimal =
   let rate = decimal (interestRate balance) / 100m
   balance * rate

let annualBalanceUpdate(balance: decimal): decimal =
   balance + (interest balance)

let amountToDonate(balance: decimal) (taxFreePercentage: float): int =
   if balance >= 0m then
       let taxFreeAmount = balance * ((decimal taxFreePercentage) / 100m)
       int (2.0m * taxFreeAmount)
       (* "You don't mind paying tax on the second half of the donation?" *)
   else
       0

File: csharp/interest-is-interesting/InterestIsInteresting.cs
static class SavingsAccount
{
    public static float InterestRate(decimal balance)
    {
        if (balance < 0) { return 3.213f; }
        else if (balance < 1_000m) { return 0.5f; }
        else if (balance < 5_000m) { return 1.621f; }
        return 2.475f;
    }

    public static decimal Interest(decimal balance)
        => balance * (decimal)InterestRate(balance) / 100.0m;

    public static decimal AnnualBalanceUpdate(decimal balance)
        => balance + Interest(balance);

    public static int YearsBeforeDesiredBalance(decimal balance, decimal targetBalance)
    {
        int result = 0;
        if (balance >= targetBalance) { result = 0; }
        else
        {
            do
            {
                balance = AnnualBalanceUpdate(balance);
                result += 1;
            } while (balance < targetBalance);
        }
        return result;
    }
}
